python/lsst/pipe/tasks/warpAndPsfMatch.py

```diff
@@ -79,44 +79,62 @@ def run(self, exposure, wcs, modelPsf=None, maxBBox=None, destBBox=None):
-            processed exposure
+            warped exposure
+        exposurePsfMatched : :cpp:class: `lsst::afw::image::Exposure`
+            warped and psf-Matched temporary exposure
         """
+        if makePsfMatched and modelPsf is None:
+            raise RuntimeError("makePsfMatched=True, but no model PSF was provided")
+
+        if not makePsfMatched and not makeDirect:
+            self.log.warn("Neither makeDirect nor makePsfMatched requested")
+
         if self.config.matchThenWarp:
             # Legacy order of operations:
             # PSF-matching is performed before warping, which is incorrect.
             # a position-dependent warping (as is used in the general case) will
             # re-introduce a position-dependent PSF.
-            if modelPsf is not None:
-                exposure = self.psfMatch.run(exposure, modelPsf).psfMatchedExposure
-            with self.timer("warp"):
-                exposure = self.warper.warpExposure(wcs, exposure, maxBBox=maxBBox, destBBox=destBBox)
+            if makePsfMatched:
+                exposurePsfMatched = self.psfMatch.run(exposure, modelPsf).psfMatchedExposure
+                with self.timer("warp"):
+                    exposurePsfMatched = self.warper.warpExposure(wcs, exposurePsfMatched,
+                                                                  maxBBox=maxBBox, destBBox=destBBox)
+            else:
+                exposurePsfMatched = None
+
+            if makeDirect:
+                # also make an unmatched temp exp
+                with self.timer("warp"):
+                    exposure = self.warper.warpExposure(wcs, exposure, maxBBox=maxBBox, destBBox=destBBox)
+            else:
+                exposure = None
+
         else:
-            if modelPsf is not None:
-                # Warp PSF before overwriting exposure
-                xyTransform = afwImage.XYTransformFromWcsPair(wcs, exposure.getWcs())
-                psfWarped = WarpedPsf(exposure.getPsf(), xyTransform)
-
-                if maxBBox is not None:
-                    # grow warped region to provide sufficient area for PSF-matching
-                    pixToGrow = 2 * max(self.config.psfMatch.kernel.active.sizeCellX,
-                                        self.config.psfMatch.kernel.active.sizeCellY)
-                    # replace with copy
-                    maxBBox = afwGeom.Box2I(maxBBox)
-                    maxBBox.grow(pixToGrow)
+            # Warp PSF before overwriting exposure
+            xyTransform = afwImage.XYTransformFromWcsPair(wcs, exposure.getWcs())
+            psfWarped = WarpedPsf(exposure.getPsf(), xyTransform)
+
+            if makePsfMatched and maxBBox is not None:
+                # grow warped region to provide sufficient area for PSF-matching
+                pixToGrow = 2 * max(self.config.psfMatch.kernel.active.sizeCellX,
+                                    self.config.psfMatch.kernel.active.sizeCellY)
```

Comment:
That's digging quite deep into the configuration tree!  You're crossing two boundaries of encapsulation, which makes the code fragile.  I think `self.psfMatch` needs a method that will return the required buffer.  That, in turn, should defer to the kernel rather than digging into the kernel's configuration.